### PR TITLE
Add a catch specific beatmap spinner gap check

### PR DESCRIPTION
**Pull request by Darius-Wattimena**
Added a first beatmap check for osu!catch, supporting all spinner (banana shower) gaps specified in the [osu!catch difficulty specific RC](https://osu.ppy.sh/wiki/en/Ranking_Criteria/osu%21catch#difficulty-specific). The spinner gap values have not been changed in the RC for years now so it should be safe to add this to the new AiMod checks (at least not in the time which I have been BN/NAT).

| Difficulty | Allowed Spinner Gap Start | Allowed Spinner Gap End |
| --- | --- | ---|
| Cup | 250 ms | 250 ms |
| Salad | 250 ms | 250 ms |
| Platter | 125 ms | 250 ms |
| Rain | 125 ms | 125 ms |
| Overdose | 62 ms | 125 ms |

Note: consecutive spinners are allowed and are an exception here with the rule.

Most likely #12091 would be the related issue, although no osu!catch specific checks have been specified in it yet.

### CODE REVIEW

File: osu.Game.Rulesets.Catch/Edit/Checks/CheckBananaShowerGap.cs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+using osu.Game.Rulesets.Objects;
+
+namespace osu.Game.Rulesets.Catch.Edit.Checks
+{
+    /// <summary>
+    /// Check the spinner/banana shower gaps specified in the osu!catch difficulty specific ranking criteria.
+    /// </summary>
+    public class CheckBananaShowerGap : ICheck
+    {
+        private static readonly int[] spinner_start_delta_threshold = { 250, 250, 125, 125, 62, 62 };
```
Comment (ekrctb) [inline osu.Game.Rulesets.Catch/Edit/Checks/CheckBananaShowerGap.cs:18]: Is it correct to apply the ranking criteria of `Expert` (Overdose) for an `ExpertPlus` beatmap?

**Review by ekrctb — DISMISSED** [review]
A confirmation is needed, otherwise looks good to me

File: osu.Game.Rulesets.Catch/Edit/Checks/CheckBananaShowerGap.cs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        private static readonly int[] spinner_end_delta_threshold = { 250, 250, 250, 125, 125, 125 };
```
Comment (bdach) [inline osu.Game.Rulesets.Catch/Edit/Checks/CheckBananaShowerGap.cs:19]: It's a nitpick but I don't really like the way this is written, kind of hard to tell what I'm looking at. Something like the following would work better, I think:
```csharp
private static readonly Dictionary<DifficultyRating, (int gapBefore, int gapAfter)> = new Dictionary<DifficultyRating, (int, int)>
{
    [DifficultyRating.Easy] = (250, 250),
    [DifficultyRating.Normal] = (250, 250),
    [DifficultyRating.Hard] = (125, 250),
    // ...
}
```

**Review by bdach — COMMENTED** [review]
Looks mostly alright

File: osu.Game.Rulesets.Catch/Edit/Checks/CheckBananaShowerGap.cs (hunk 2 above)
Comment (Darius-Wattimena) [reply osu.Game.Rulesets.Catch/Edit/Checks/CheckBananaShowerGap.cs]: Should be better now, thanks for the comment, this is better for sure.

**Review by bdach — APPROVED** [review]

**Review by ekrctb — APPROVED** [review]

**Pull request merged**